adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,551 @@
+[code]#cudaStreamSynchronize()# blocks work from being performed until all
+threads on the device has been completed.
+[code]#__syncthreads()# waits for
+all threads within a thread block to reach the same point. So 
+[code]#cudaStreamSynchronize()# is similar to queue.wait(), buffer
+destruction, and other host-device synchronization events within SYCL.
+[code]#__syncthreads()# synchronizes the threads within a thread block which
+is analogous to the work-group barrier.
+
+CUDA's warp concept has no SYCL equivalent. If a user were to write 
+warp aware code it would be non-generic SYCL code and specific to the 
+CUDA backend.
+
+CUDA allows for more detailed thread and memory management through 
+Cooperative Groups. Cooperative Groups allow for synchronizing at the 
+grid level and organizing subgroups in sizes smaller than a warp. 
+Cooperative Groups do not have an equivalent within SYCL 2020 and are 
+not yet supported.
+
+==== Work Item Mapping
+
+The SYCL specification specifies that work-items must be arranged in a row major
+fashion, making work-items with ids `(a, b, c)` and `(a, b, c+1)` adjacent.
+
+In native CUDA, work-items are arranged in a column major fashion, making 
+work-items with ids `(a, b, c)` and `(a+1, b, c)` adjacent.
+
+In order for a given SYCL implementation's CUDA backend to conform to the SYCL
+specification, the implementation must map the row major ordering of SYCL to the
+column major ordering specific to the CUDA backend. The underlying column major
+ordering of work-items in CUDA is therefore should be transparent to the user,
+however, could still be observable in kernel code which uses backend
+interoperability.
+
+[[table.cuda.CUDA_features_to_SYCL]]
+.CUDA execution features with their corresponding SYCL features
+[width="100%",options="header",cols="50%,50%"]
+|====
+| [code]#SYCL#                                                       | [code]#CUDA#
+| [code]#Compute unit#                                               | [code]#Streaming multiprocessor#
+| [code]#nd_range#                                                   | [code]#grid#
+| [code]#work-group#                                                 | [code]#Thread block#
+| [code]#sub-group#                                                  | [code]#thread_block_tile#
+| [code]#work-item#                                                  | [code]#Thread#
+| [code]#SYCL nd_item synchronization#                               | [code]#cudaStreamSynchronize#
+| [code]#work-group barrier#                                         | [code]#__syncthreads#
+|====
+
+[[sec::programming_interface]]
+== Programming Interface
+
+[[table.cuda.get_error_code]]
+.Free function for querying whether an event has a native CUevent
+[width="100%",options="header",cols="40%,60%"]
+|====
+| Function | Description
+| [code]#bool cuda::has_native_event()# | Returns true if the SYCL [code]#event# has a native [code]#CUevent# associated with it.
+|====
```

Comment:
How does this relate to `sycl::error_category_for` (see #164)?  If the idea is that each backend interop spec will provide a `get_error_code` function like this, then can we remove the ill-defined `sycl::error_category_for` from the core spec?

It would be very good to resolve #164 because we deferred any CTS testing of `sycl::error_category_for` due to its unclear specification.

Reply:
Sorry I missed your comment here. Are you referring to `cuda::get_error_code(exception &)` and how that relates to `sycl::error_category_for`?

Yeah, I think the consensus was that we should move towards per-backend `get_error_code` functions, so then we can remove `sycl::error_category_for`. Though I'm not sure if it was decided exactly what we should do with `sycl::error_category_for`, whether it's removed immediately from SYCL 2020 or deprecated, I think there is justification for removing immediately it as it's been found to not work as it was intended.